c2rust-analyze/src/main.rs

```diff
@@ -449,13 +449,121 @@ fn run(tcx: TyCtxt) {
+                        Place<'tcx>,
+                    )>| {
+                        let deref_ty = void_ptr
+                            .ty(acx.local_decls, acx.tcx())
+                            .ty
+                            .builtin_deref(true)
+                            .unwrap()
+                            .ty
+                            .kind();
+                        assert_matches!(deref_ty, TyKind::Adt(adt, _) => {
+                            assert_eq!(tcx.def_path(adt.did()).data[0].to_string(), "ffi");
+                            assert_eq!(tcx.item_name(adt.did()).as_str(), "c_void");
+                        });
+
+                        if let Some((casted, castee)) = find_cast() {
+                            if castee == *void_ptr {
+                                // This is a special case for types being casted from *c_void to a pointer
+                                // to some other type, e.g. `let foo = malloc(..) as *mut Foo;`
+                                // carry over the pointer id of *c_void, but match the pointer ids
+                                // of the casted-to-type for the rest
+                                acx.c_void_casts.0.insert(castee, casted);
+                            }
+                        }
+                    };
+
+                    let find_last_cast_curr_block = || {
+                        bb_data.statements.iter().rev().find_map(|s| match &s.kind {
+                            StatementKind::Assign(x) => match &x.1 {
+                                Rvalue::Cast(_, op, _) => Some((
+                                    op.place().expect(
+                                        "Rvalue::Cast is not Operant::Constant so it has a Place",
+                                    ),
+                                    x.0,
+                                )),
+                                _ => None,
+                            },
+                            _ => None,
+                        })
+                    };
+
+                    let find_first_cast_succ_block = || {
+                        let successor = target.unwrap();
+                        mir.basic_blocks()[successor]
+                            .statements
+                            .iter()
+                            .find_map(|s| match &s.kind {
+                                StatementKind::Assign(x) => match &x.1 {
+                                    Rvalue::Cast(_, op, _) => Some((
+                                        x.0,
+                                        op.place().expect(
+                                            "Rvalue::Cast is not Operant::Constant so it has a Place",
+                                        ),
+                                    )),
+                                    _ => None,
+                                },
+                                _ => None,
+                            })
+                    };
```

Comment:
The two `find_map` closures are the same and can be refactored out.  I don't say this just to reduce repetition, but there are often little differences that sneak into one of the versions when they're separate.  Like there was for `CVoidCasts::contains` until it was refactored into a single version.

Reply:
They are actually slightly different; they return tuples in reverse order of one another. But I will look for a way to factor it out anyway

Reply:
60fe64d8

Reply:
Thanks, that's a lot clearer now.  Much easier to see they're flipped in order now.